&nbsp;
<div style="text-align: center;">

## <span style="color: black;">Website SuperClassics - Europe Classic Cars</span>

&nbsp;

![SuperClassics Logo](superclassics_logo.png)

&nbsp;

### <span style="color: black;">Data Cleaning</span>

#### <span style="color: #003E99;">Variables: Brand, Model, Submodel, Year, Origin</span>

#### <span style="color: #003E99;">2272 Cars</span>
    
</div>

&nbsp;

&nbsp;


## Index

- [<font color='black'> 1. Packages installation</font>](#1)



- [<font color='black'> 2. Data importation</font>](#2)



- [<font color='black'> 3. Data exploration</font>](#3)
    - [<font color='black'> 3.1. Missing values</font>](#3.1)
    - [<font color='black'> 3.2. Most frequent itens</font>](#3.2)
    - [<font color='black'> 3.3. Brand variable</font>](#3.3)
    - [<font color='black'> 3.4. Model variable</font>](#3.4)
    - [<font color='black'> 3.5. Submodel variable</font>](#3.5)
    - [<font color='black'> 3.6. Year variable</font>](#3.6)
    - [<font color='black'> 3.7. Origin variable</font>](#3.7)
    
    
- [<font color='black'> 5. Data cleaning</font>](#5)
    - [<font color='black'> 5.1. Brand variable</font>](#5.1)
    - [<font color='black'> 5.2. Model variable</font>](#5.2)
    - [<font color='black'> 5.3. Submodel variable</font>](#5.3)
    - [<font color='black'> 5.4. Year variable</font>](#5.4)
    - [<font color='black'> 5.5. Origin variable</font>](#5.5)
    
    
- [<font color='black'> 6. Data exportation</font>](#6)

## 1. Packages installation
<a id='1'>

In [1]:
#!!!INSTALL FIIRST TIME!!!
#!pip3 install unicode

In [117]:
#dataframes
import pandas as pd

# compativel com linguagem ASCII - unicode
from unidecode import unidecode 

#Expressões regulares
import re


## 2. Data importation
<a id='2'>

In [118]:
db = pd.read_csv('Manufacturers_rawdata.csv', sep=';')

## 3. Data exploration
<a id='3'>

In [119]:
db.head()

Brand                  Model Submodel         Year Origin
0    16  Hurricane Preselector      NaN  1946 - 1948    NaN
1    18              Hurricane      NaN  1948 - 1953    NaN
2    18  Hurricane Preselector      NaN  1948 - 1953    NaN
3    18              Limousine      NaN  1950 - 1951    NaN
4    18                Whitley      NaN  1948 - 1953    NaN

<span style="color:blue">
    
- Here we can observe that there is incorrect brands, envolving numbers (eg. 16, 18,...)

In [120]:
db.tail()

Brand   Model   Submodel         Year   Origin
2267     Volvo   P1800         ES  1971 - 1973   Sweden
2268     Volvo  P 1800          S  1968 - 1969   Sweden
2269     Volvo  P 1800          S  1963 - 1968   Sweden
2270  Wartburg     353  W Tourist  1975 - 1989  Germany
2271   Zündapp   Janus        NaN  1957 - 1958  Germany

### 3.1. Missing values
<a id='3.1'>

In [121]:
db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2272 entries, 0 to 2271
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Brand     2271 non-null   object
 1   Model     2262 non-null   object
 2   Submodel  2000 non-null   object
 3   Year      2272 non-null   object
 4   Origin    2258 non-null   object
dtypes: object(5)
memory usage: 88.9+ KB


<span style="color:blue">
    
- All variables with **data type "object"** , which is a categorical type (Year variable should be numeric?)


    
| Variable     | NaN |
|:-----------------|-----------------------|
| Brand       | 1                 |
| Model       | 10                |
| Submodel    | 272               |
| Year        | 0                |
| Origin      | 14                |


<span style="color:blue">
    
- Submodel variable has a lot of NaN values

### 3.2. Most frequent itens
<a id='3.2'>

In [122]:
db.describe()

Brand Model Submodel         Year Origin
count      2271  2262     2000         2272   2258
unique       94   480     1338          428     17
top     Citroën    CX       GT  1974 - 1975  Italy
freq        335    97       23           42    780

<span style="color:blue">
    
- **Brand** --> 94 unique classic car brands, citroen as most popular brand with 335 different classic car models
    
- **Model** --> 480 unique classic car models, CX is the most frequent model, with 97 models named
    
- **Submodel** --> 1338 unique classic car submodels, GT is the most frequent submodel, with 23 submodels named
    
- **Year** --> the most frequent manufactured year is between 1974 and 1975, with 42 classic cars manufactured
    
- **Origin** --> Italy is the country with the most classic cars manufactures, with a total of 780 cars

### 3.3. Brand variable
<a id='3.3'>

In [123]:
unique_brand = db['Brand'].unique().tolist()
print(unique_brand)

['16', '18', '206', '208', '246', '308', 'Abarth', 'Adler', 'Alfa Romeo', 'Allard', 'ALPINA', 'Alpine', 'Alta', 'Alvis', 'Amphicar', 'Apollo', 'Armstrong Siddeley', 'Arnolt-Bristol', 'Ascort', 'Aston Martin', 'Audi', 'Austin', 'Austin-Healey', 'Authi', 'Auto Union', 'Autobianchi', 'Bentley', 'Bitter', 'Bizzarini', 'BMW', 'Borgward', 'Bristol', 'Bugatti', 'Buick', 'Cadillac', 'Chevrolet', 'Citroëm', 'Citroën', 'Daf', 'DAF', 'Daimler', 'Datsun', 'De Tomaso', 'Delage', 'Delahaye', 'DeLorean', 'DKW', 'Donkervoort', 'Excalibur', 'Facel Vega', 'Ferrari', 'Fiat', 'Ford', nan, 'Fuldamobil', 'Glas', 'Goggomobil', 'Heinkel', 'Honda', 'Innocenti', 'Iso Rivolta', 'Jaguar', 'Jeep', 'Jensen', 'Lamborghini', 'Lancia', 'Land Rover', 'Landrover', 'Lloyd', 'Lotus', 'Maserati', 'Matra', 'Mercedes-Benz', 'Messerschmitt', 'MG', 'Morris', 'NSU', 'Opel', 'Peugeot', 'Porsche', 'Renault', 'Rolls-Royce', 'Rover', 'Saab', 'Simca', 'Simca 9', 'Sunbeam', 'Talbot', 'Trabant', 'Triumph', 'TVR', 'Volkswagen', 'Volvo'

<span style='color:blue'> 
    
**To solve:**
    
- remove accents and hifens 
    
- citroem to citroen , DAF to Daf, Simca 9 to Simca
    
**Values to explore:**
    
- 16, 18, 206, 208, 246, 308, nan, Simca 9 

### 3.4. Model variable
<a id='3.4'>

In [124]:
unique_model = db['Model'].unique().tolist()
unique_model_ordem = sorted(unique_model, key=lambda x:(isinstance(x,str),x))
print(unique_model_ordem)

[nan, '1 281 100', '1.1', '100', '100-4', '100-4 BN1', '100-6', '1000', '1004', '100M', '107', '109', '1200', '1204', '122', '124', '124 Estate', '124 Special', '124 Special T', '124 Sport Coupe 1400', '124 Sport Coupe 1600', '124 Sport Coupé', '124 Sport Spider 1400', '124 Sport Spider 1600', '124 Sport Spider 1800', '126', '126 L', '126 Sunroof', '127', '128', '128 1100 Special', '128 1300 Special', '128 Kombi 1100', '128 Rally', '128 Sport Coupé S', '128 Sport Coupé SL', '130', '130 2.9 V6', '130 3.2 V6', '130 Coupé 3.2 V6', '1300', '1304', '131', '132', '135', '145', '146', '1500', '1600', '164', '166 Inter', '17 M', '170', '1700', '1750 Spider', '190 SL', '1900', '2 Liter', '2.8', '200', '2000', '2002', '204', '208', '220', '2200', '228', '230', '240', '242', '250', '2500', '2500 M', '260', '2600', '280', '2800', '288', '3.0', '3.3', '30', '300', '300 SL', '3000', '3000 Mk I', '3000 Mk II', '3000 Mk III', '303', '308', '309', '31', '32', '320', '3200', '326', '327', '328', '33', '

### 3.5. Submodel variable
<a id='3.5'>

In [125]:
unique_submodel = db['Submodel'].unique().tolist()
unique_submodel_ordem = sorted(unique_submodel, key=lambda x:(isinstance(x,str),x))
print(unique_submodel_ordem)

[nan, '"Gullwing"', '(Tipo 116)', '/4', '1.0 Harmonie', '1.0 Teen', '1.0 Teen Pop', '1.0 Teen Spot', '1.0 Teen Tonic', '1.0 Top', '1.1 Fashion', '1.1 First', '1.1 GL', '1.1 GLS', '1.1 Hit', '1.1 LS', '1.1 TRS Classic', '1.2', '1.2 L', '1.3', '1.3  Tipo 116 2a', '1.3 GLS', '1.3 L', '1.3 LSi', '1.3 Milano', '1.3 Rallye 2', '1.3 SC', '1.3 SL', '1.3 Super', '1.3 ti', '1.3i', '1.3i Basic', '1.3i Catalyst', '1.3i Catalyst College', '1.3i College', '1.3i Noblesse', '1.4', '1.4 16V T.Spark', '1.4 16V T.Spark Edizione Sportiva', '1.4 16V T.Spark Junior', '1.4 16V T.Spark L', '1.4 16V T.Spark Limited Edition 500', '1.4 4x4', '1.4 Avantage', '1.4 Blue Note', '1.4 D First', '1.4 D TGD', '1.4 D Teen', '1.4 Easy', '1.4 Fashion', '1.4 GL', '1.4 GT', '1.4 GTI', '1.4 Harmonie', '1.4 I.E. Imola', '1.4 I.E. L', '1.4 L', '1.4 Premiere', '1.4 Prestige', '1.4 Reflex', '1.4 TZX Tecnic', '1.4 Tonic', '1.4 X', '1.5', '1.5 4x4', '1.5 D Teen', '1.5 D Top', '1.5 GT', '1.5 LSi', '1.5 Milano', '1.5 Quadrifoglio', '

### 3.6. Year variable
<a id='3.6'>

In [126]:
unique_year = db['Year'].unique().tolist()
print(unique_year)

['1946 - 1948', '1948 - 1953', '1950 - 1951', '1968 - 1969', '1966', '1975 - 1976', '1969 - 1970', '1970 - 1971', '1971 - 1974', '1972 - 1974', '1973 - 1975', '1973 - 1976', '1962 - 1963', '1959 - 1961', '1969 - 1971', '1976 - 1978', '1952', '1960 - 1962', '1971 - 1976', '1976 - 1985', '1985 - 1986', '1983 - 1985', '1981 - 1983', '1962  - 1963', '1964 - 1965', '1963 - 1964', '1938 - 1940', '1936 - 1941', '1937 - 1940', '1967 - 1969', '1954 - 1955', '1975-1975', '1962 - 1965', '1986 - 1989', '1979 - 1983', '1947 - 1950', '1984 - 1986', '1972 - 1980', '1980 - 1981', '1976 -1979', '1974 - 1975', '1976 - 1980', '1979 - 1980', '1966 - 1974', '1965 - 1966', '1971 - 1977', '1992 - 1993', '1994 - 1996', '1997 - 1999', '1999 - 2000', '2000', '1998 - 1999', '1997 - 1998', '1995 - 1996', '1996', '1995 - 1997', '1996 - 1997', '1996 - 1998', '1987 - 1993', '1991 - 1993', '1988 - 1993', '1993 - 1994', '1989 - 1993', '1994 - 1997', '1990 - 1993', '1993 - 1997', '1954', '1970 - 1976', '1983 - 1986', '

<span style='color:blue'> 

**To solve:**
    
- Transformation into two new varibles: 'start_year' and 'end_year'
    
- Cases out of format XXXX - XXXX


### 3.7. Origin variable
<a id='3.7'>

In [127]:
unique_origin = db['Origin'].unique().tolist()
print(unique_origin)

[nan, 'Italy', 'italy', 'Germany', 'United Kingdom', 'France', 'Greece', 'United States', 'Australia', 'Spain', '1977', 'Netherlands', 'Japan', 'Ireland', 'Germany / United Kingdom', 'Germany / United Kingdom up to 1976', 'Sweden', 'Zweden']


<span style='color:blue'> 
    
**To do:**
    
- remove uppercase letters?
    
    
- remove 1977, gemany/united kingdom, germany/united kingdom up to 1976?
    
    
(mantain all countries, even outside europe)

## 5. Data cleaning
<a id='5'>

In [128]:
db_clean = pd.DataFrame(db)

### 5.1. Brand variable
<a id='5.1'>

#### Remove accents

In [129]:
#to string to use unicode
db_clean['Brand'] = db_clean['Brand'].apply(lambda x: str(x))

db_clean['Brand'] = db_clean['Brand'].apply(unidecode)

#### Transformation of hifen into space + uppercase

In [130]:
db_clean['Brand'] = db_clean['Brand'].str.replace('-', ' ').str.title()

In [131]:
unique_brand_clean = db_clean['Brand'].unique().tolist()
print(unique_brand_clean)

['16', '18', '206', '208', '246', '308', 'Abarth', 'Adler', 'Alfa Romeo', 'Allard', 'Alpina', 'Alpine', 'Alta', 'Alvis', 'Amphicar', 'Apollo', 'Armstrong Siddeley', 'Arnolt Bristol', 'Ascort', 'Aston Martin', 'Audi', 'Austin', 'Austin Healey', 'Authi', 'Auto Union', 'Autobianchi', 'Bentley', 'Bitter', 'Bizzarini', 'Bmw', 'Borgward', 'Bristol', 'Bugatti', 'Buick', 'Cadillac', 'Chevrolet', 'Citroem', 'Citroen', 'Daf', 'Daimler', 'Datsun', 'De Tomaso', 'Delage', 'Delahaye', 'Delorean', 'Dkw', 'Donkervoort', 'Excalibur', 'Facel Vega', 'Ferrari', 'Fiat', 'Ford', 'Nan', 'Fuldamobil', 'Glas', 'Goggomobil', 'Heinkel', 'Honda', 'Innocenti', 'Iso Rivolta', 'Jaguar', 'Jeep', 'Jensen', 'Lamborghini', 'Lancia', 'Land Rover', 'Landrover', 'Lloyd', 'Lotus', 'Maserati', 'Matra', 'Mercedes Benz', 'Messerschmitt', 'Mg', 'Morris', 'Nsu', 'Opel', 'Peugeot', 'Porsche', 'Renault', 'Rolls Royce', 'Rover', 'Saab', 'Simca', 'Simca 9', 'Sunbeam', 'Talbot', 'Trabant', 'Triumph', 'Tvr', 'Volkswagen', 'Volvo', 'Wa

####  Explore values: 16, 18, 206, 208, 246, 308, nan, Simca 9

In [132]:
print(db_clean[db_clean['Brand'] == '16'])
print(db_clean[db_clean['Brand'] == '18'])
print(db_clean[db_clean['Brand'] == '206'])
print(db_clean[db_clean['Brand'] == '208'])
print(db_clean[db_clean['Brand'] == '246'])
print(db_clean[db_clean['Brand'] == '308'])
print(db_clean[db_clean['Brand'] == 'nan'])
print(db_clean[db_clean['Brand'] == 'Simca 9'])

  Brand                  Model Submodel         Year Origin
0    16  Hurricane Preselector      NaN  1946 - 1948    NaN
  Brand                  Model Submodel         Year Origin
1    18              Hurricane      NaN  1948 - 1953    NaN
2    18  Hurricane Preselector      NaN  1948 - 1953    NaN
3    18              Limousine      NaN  1950 - 1951    NaN
4    18                Whitley      NaN  1948 - 1953    NaN
  Brand Model Submodel         Year Origin
5   206   NaN       GT  1968 - 1969  Italy
6   206   NaN       SP         1966  Italy
  Brand Model Submodel         Year Origin
7   208   NaN      GT4  1975 - 1976  Italy
   Brand Model Submodel         Year Origin
8    246   NaN       GT  1969 - 1970  Italy
9    246   NaN       GT  1970 - 1971  Italy
10   246   NaN       GT  1971 - 1974  Italy
11   246   NaN      GTS  1972 - 1974  Italy
   Brand Model Submodel         Year Origin
12   308   NaN      GT4  1973 - 1975  Italy
13   308   NaN      GT4  1973 - 1976  Italy
Empty DataFra

In [133]:
db_clean.loc[[0,1,2,3,4,5,6,7,8,9,10,11,12,13,1270, 2229]]

Brand                  Model Submodel         Year  Origin
0          16  Hurricane Preselector      NaN  1946 - 1948     NaN
1          18              Hurricane      NaN  1948 - 1953     NaN
2          18  Hurricane Preselector      NaN  1948 - 1953     NaN
3          18              Limousine      NaN  1950 - 1951     NaN
4          18                Whitley      NaN  1948 - 1953     NaN
5         206                    NaN       GT  1968 - 1969   Italy
6         206                    NaN       SP         1966   Italy
7         208                    NaN      GT4  1975 - 1976   Italy
8         246                    NaN       GT  1969 - 1970   Italy
9         246                    NaN       GT  1970 - 1971   Italy
10        246                    NaN       GT  1971 - 1974   Italy
11        246                    NaN      GTS  1972 - 1974   Italy
12        308                    NaN      GT4  1973 - 1975   Italy
13        308                    NaN      GT4  1973 - 1976   Italy
1270      Nan                    280     ZX T  1979 - 1981   Japan
2229  Simca 9                      9      NaN  1952 - 1954  France

<span style="color:blue">
    
**Case 16, 18** - No brand or model information for the car; hurricane preselector is an automatic transmission, with this information, it is not possible to recognize the brand and model of the car - Remove **records 0, 1, 2, 3, 4**
    
&nbsp;    
       

**Case 206** - Ferrari Dino 206 GT - Replace brand **206** with **Ferrari**, model **NaN** with **Dino 206**, in **record 5**
    
**Case 206** - Ferrari 206 SP - Replace brand **206** with **Ferrari**, model **NaN** with **206**, in **record 6**
    
&nbsp;
    
**Case 208** - Ferrari 208 GT4 - Replace brand **208** with **Ferrari**, model **NaN** with **208**, in **record 7**
    
&nbsp;
    
**Case 246** - Ferrari Dino 246 GT - Replace brand **246** with **Ferrari**, model **NaN** with **Dino 246**, in **record 8**
    
    
**Case 246** - Ferrari Dino 246 GT - Replace brand **246** with **Ferrari**, model **NaN** with **Dino 246**, in **record 9**
    
**Case 246** - Ferrari Dino 246 GT - Replace brand **246** with **Ferrari**, model **NaN** with **Dino 246**, in **record 10**
    
**Case 246** - Ferrari Dino 246 GTS - Replace brand **246** with **Ferrari**, model **NaN** with **Dino 246**, in **record 10** 
    
(Case 246, the model was extended over the years)
    
    
&nbsp;
    
    
**Case 308** - Ferrari 308 GT4 - Replace brand **308** with **Ferrari**, model **NaN** with **308**, in **record 12**
        
**Case 308** - Ferrari 308 GT4 - Replace brand **308** with **Ferrari**, model **NaN** with **308**, in **record 13**
    
(Case 308, model extended over the years)

&nbsp;
    
**Case nan** - Datsun 280 ZX Turbo - Replace brand **nan** with **Datsun**, in **record 1270** 
    
&nbsp;
    
**Case Simca 9** - Simca 9 - Replace brand **Simca 9** with **Simca**, in **record 2229**
    
(Case Simca 9 has no submodel)



In [134]:
# Values replacement

db_clean.loc[5, 'Brand'] = 'Ferrari'
db_clean.loc[5, 'Model'] = 'Dino 206'

db_clean.loc[6, 'Brand'] = 'Ferrari'
db_clean.loc[6, 'Model'] = '206'

db_clean.loc[7, 'Brand'] = 'Ferrari'
db_clean.loc[7, 'Model'] = '208'

db_clean.loc[8, 'Brand'] = 'Ferrari'
db_clean.loc[8, 'Model'] = 'Dino 246'

db_clean.loc[9, 'Brand'] = 'Ferrari'
db_clean.loc[9, 'Model'] = 'Dino 246'

db_clean.loc[10, 'Brand'] = 'Ferrari'
db_clean.loc[10, 'Model'] = 'Dino 246'

db_clean.loc[11, 'Brand'] = 'Ferrari'
db_clean.loc[11, 'Model'] = 'Dino 246'

db_clean.loc[12, 'Brand'] = 'Ferrari'
db_clean.loc[12, 'Model'] = '308'

db_clean.loc[13, 'Brand'] = 'Ferrari'
db_clean.loc[13, 'Model'] = '308'

db_clean.loc[1270, 'Brand'] = 'Datsun'

db_clean.loc[2229, 'Brand'] = 'Simca'


In [135]:
#verification

db_clean.loc[[5,6,7,8,9,10,11,12,13,1270, 2229]]


Brand     Model Submodel         Year  Origin
5     Ferrari  Dino 206       GT  1968 - 1969   Italy
6     Ferrari       206       SP         1966   Italy
7     Ferrari       208      GT4  1975 - 1976   Italy
8     Ferrari  Dino 246       GT  1969 - 1970   Italy
9     Ferrari  Dino 246       GT  1970 - 1971   Italy
10    Ferrari  Dino 246       GT  1971 - 1974   Italy
11    Ferrari  Dino 246      GTS  1972 - 1974   Italy
12    Ferrari       308      GT4  1973 - 1975   Italy
13    Ferrari       308      GT4  1973 - 1976   Italy
1270   Datsun       280     ZX T  1979 - 1981   Japan
2229    Simca         9      NaN  1952 - 1954  France

In [136]:
# removal of values: 16,18

db_clean = db_clean.drop(0)
db_clean = db_clean.drop(1)
db_clean = db_clean.drop(2)
db_clean = db_clean.drop(3)
db_clean = db_clean.drop(4)

#### Transformation of Citroem into Citroen, DAF into Daf (...), Simca 9 into Simca

In [137]:
db_clean['Brand'] = db_clean['Brand'].replace('Citroem', 'Citroen')
db_clean['Brand'] = db_clean['Brand'].replace('Daf', 'DAF')
db_clean['Brand'] = db_clean['Brand'].replace('Bmw', 'BMW')
db_clean['Brand'] = db_clean['Brand'].replace('Dkw', 'DKW')
db_clean['Brand'] = db_clean['Brand'].replace('Mg', 'MG')
db_clean['Brand'] = db_clean['Brand'].replace('Nsu', 'NSU')
db_clean['Brand'] = db_clean['Brand'].replace('Simca 9', 'Simca')
db_clean['Brand'] = db_clean['Brand'].replace('Tvr', 'TVR')


In [138]:
# verification

unique_brand_clean = db_clean['Brand'].unique().tolist()
print(unique_brand_clean)


['Ferrari', 'Abarth', 'Adler', 'Alfa Romeo', 'Allard', 'Alpina', 'Alpine', 'Alta', 'Alvis', 'Amphicar', 'Apollo', 'Armstrong Siddeley', 'Arnolt Bristol', 'Ascort', 'Aston Martin', 'Audi', 'Austin', 'Austin Healey', 'Authi', 'Auto Union', 'Autobianchi', 'Bentley', 'Bitter', 'Bizzarini', 'BMW', 'Borgward', 'Bristol', 'Bugatti', 'Buick', 'Cadillac', 'Chevrolet', 'Citroen', 'DAF', 'Daimler', 'Datsun', 'De Tomaso', 'Delage', 'Delahaye', 'Delorean', 'DKW', 'Donkervoort', 'Excalibur', 'Facel Vega', 'Fiat', 'Ford', 'Fuldamobil', 'Glas', 'Goggomobil', 'Heinkel', 'Honda', 'Innocenti', 'Iso Rivolta', 'Jaguar', 'Jeep', 'Jensen', 'Lamborghini', 'Lancia', 'Land Rover', 'Landrover', 'Lloyd', 'Lotus', 'Maserati', 'Matra', 'Mercedes Benz', 'Messerschmitt', 'MG', 'Morris', 'NSU', 'Opel', 'Peugeot', 'Porsche', 'Renault', 'Rolls Royce', 'Rover', 'Saab', 'Simca', 'Sunbeam', 'Talbot', 'Trabant', 'Triumph', 'TVR', 'Volkswagen', 'Volvo', 'Wartburg', 'Zundapp']


### 5.2. Year variable
<a id='5.2'>

### Transformação em ano inicio e fim

In [139]:

# clean and transform year variable into two new variables 'start_year'and 'end_year'
def clean_and_split_years(year_str):
    
    # Remove all non numeric caracters,except hifen
    year_str = re.sub(r'[^\d-]', '', year_str)
    
    # Divide by hifen
    years = year_str.split('-')
    
    # If only one year, add empty second year 
    if len(years) == 1:
        years.append('')
    
    return years

# apply function
db_clean['Start_Year'], db_clean['End_Year'] = zip(*db_clean['Year'].map(clean_and_split_years))

# remove 'year' column
db_clean = db_clean.drop(columns=['Year'])


In [140]:
#Verification

db_clean

Brand     Model   Submodel   Origin Start_Year End_Year
5      Ferrari  Dino 206         GT    Italy       1968     1969
6      Ferrari       206         SP    Italy       1966         
7      Ferrari       208        GT4    Italy       1975     1976
8      Ferrari  Dino 246         GT    Italy       1969     1970
9      Ferrari  Dino 246         GT    Italy       1970     1971
...        ...       ...        ...      ...        ...      ...
2267     Volvo     P1800         ES   Sweden       1971     1973
2268     Volvo    P 1800          S   Sweden       1968     1969
2269     Volvo    P 1800          S   Sweden       1963     1968
2270  Wartburg       353  W Tourist  Germany       1975     1989
2271   Zundapp     Janus        NaN  Germany       1957     1958

[2267 rows x 6 columns]

In [141]:
# transform new variable into numeric
db_clean['Start_Year'] = pd.to_numeric(db_clean['Start_Year'])
db_clean['End_Year'] = pd.to_numeric(db_clean['End_Year'])

In [142]:
#show new data description for year variable
db_clean.describe()

Start_Year     End_Year
count  2.267000e+03  2134.000000
mean   1.930974e+04  1980.429241
std    5.833716e+05    18.692001
min    1.928000e+03  1698.000000
25%    1.969000e+03  1972.000000
50%    1.979000e+03  1982.000000
75%    1.991000e+03  1993.000000
max    1.966197e+07  2018.000000

<span style="color:blue">
    
- **'start_year variable** --> first classic car manufactured starts at 1928 and last classic car manufactured starts at 1966
    
- **'end_year variable** --> first classic car manufactured ends at ? and last classic car manufactured ends at 2018
    
    
**To do:**
    
- investigate value 1698 and years out of format

#### Investigate 1698 value in end_year

In [143]:
print(db_clean[db_clean['End_Year'] == 1698])

     Brand       Model Submodel Origin  Start_Year  End_Year
1239  Fiat   850 Coupé      NaN  Italy        1965    1698.0
1240  Fiat       850 N      NaN  Italy        1964    1698.0
1242  Fiat       850 S      NaN  Italy        1964    1698.0
1244  Fiat  850 Spider      NaN  Italy        1965    1698.0


<span style= 'color:blue'>
    
- the four fiat models had incorrect values, these were corrected below

In [144]:
# Values replacement
db_clean.loc[1239, 'End_Year'] = 1971

db_clean.loc[1240, 'End_Year'] = 1971

db_clean.loc[1242, 'Start_Year'] = 1968
db_clean.loc[1242, 'End_Year'] = 1972

db_clean.loc[1244, 'End_Year'] = 1973

In [145]:
# verification
unique_year_clean1 = db_clean['Start_Year'].unique().tolist()
print(unique_year_clean1)

[1968, 1966, 1975, 1969, 1970, 1971, 1972, 1973, 1962, 1959, 1976, 1952, 1960, 1985, 1983, 1981, 1964, 1963, 1938, 1936, 1937, 1967, 1954, 1986, 1979, 1947, 1984, 1980, 1974, 1965, 1992, 1994, 1997, 1999, 2000, 1998, 1995, 1996, 1987, 1991, 1988, 1993, 1989, 1990, 1950, 1982, 1978, 1977, 1958, 1949, 1932, 1961, 1955, 1951, 1953, 1957, 1933, 1935, 1956, 2004, 2006, 1934, 1946, 1928, 1930, 1939, 1948, 1931, 2016, 19661969, 19631967, 2001, 2003, 2002, 1945]


#### Verify 19661969 and 19631967 values

In [146]:
print(db_clean[db_clean['Start_Year'] == 19661969])
print(db_clean[db_clean['Start_Year'] == 19631967])


     Brand    Model Submodel         Origin  Start_Year  End_Year
1402  Ford  Mustang      390  United States    19661969       NaN
     Brand Model Submodel   Origin  Start_Year  End_Year
1422  Glas  1004      NaN  Germany    19631967       NaN


<span style= 'color:blue'>
    
- these models have not End_year, probably because of bad join, there were corrected below

In [147]:
db_clean.loc[1402, 'Start_Year'] = 1966
db_clean.loc[1402, 'End_Year'] = 1969

db_clean.loc[1422, 'Start_Year'] = 1963
db_clean.loc[1422, 'End_Year'] = 1967

In [148]:
# verification
unique_year_clean1 = db_clean['Start_Year'].unique().tolist()
print(unique_year_clean1)

[1968, 1966, 1975, 1969, 1970, 1971, 1972, 1973, 1962, 1959, 1976, 1952, 1960, 1985, 1983, 1981, 1964, 1963, 1938, 1936, 1937, 1967, 1954, 1986, 1979, 1947, 1984, 1980, 1974, 1965, 1992, 1994, 1997, 1999, 2000, 1998, 1995, 1996, 1987, 1991, 1988, 1993, 1989, 1990, 1950, 1982, 1978, 1977, 1958, 1949, 1932, 1961, 1955, 1951, 1953, 1957, 1933, 1935, 1956, 2004, 2006, 1934, 1946, 1928, 1930, 1939, 1948, 1931, 2016, 2001, 2003, 2002, 1945]


In [149]:
# verification
unique_year_clean2 = db_clean['End_Year'].unique().tolist()
print(unique_year_clean2)

[1969.0, nan, 1976.0, 1970.0, 1971.0, 1974.0, 1975.0, 1963.0, 1961.0, 1978.0, 1962.0, 1985.0, 1986.0, 1983.0, 1965.0, 1964.0, 1940.0, 1941.0, 1955.0, 1989.0, 1950.0, 1980.0, 1981.0, 1979.0, 1966.0, 1977.0, 1993.0, 1996.0, 1999.0, 2000.0, 1998.0, 1997.0, 1994.0, 1984.0, 1987.0, 1990.0, 1992.0, 1952.0, 1988.0, 1982.0, 1959.0, 1967.0, 1968.0, 1953.0, 1995.0, 1936.0, 1960.0, 1956.0, 1957.0, 2003.0, 1951.0, 1954.0, 2001.0, 1934.0, 1938.0, 1973.0, 2006.0, 2009.0, 2004.0, 2010.0, 1972.0, 1958.0, 1939.0, 1991.0, 1949.0, 1931.0, 1932.0, 1944.0, 1933.0, 1937.0, 1935.0, 2002.0, 2018.0, 2017.0, 1948.0, 2005.0, 2007.0]


In [150]:
x=db_clean[pd.isnull(db_clean['End_Year'])]
print(x)


           Brand   Model                             Submodel   Origin  \
6        Ferrari     206                                   SP    Italy   
19        Abarth    1500                        Coupé Biposto    italy   
57    Alfa Romeo     145    1.4 16V T.Spark Edizione Sportiva    Italy   
61    Alfa Romeo     145  1.4 16V T.Spark Limited Edition 500    Italy   
66    Alfa Romeo     145    1.6 16V T.Spark Edizione Sportiva    Italy   
...          ...     ...                                  ...      ...   
2184     Porsche     911                        Carrera Coupé  Germany   
2186     Porsche     911                          Coupe 2.0 L  Germany   
2204     Porsche     924                           Carrera GT  Germany   
2210     Porsche     944                             Coupé S2  Germany   
2227        Saab  Ursaab                                  NaN   Sweden   

      Start_Year  End_Year  
6           1966       NaN  
19          1952       NaN  
57          2000       N

#### Remove decimal unit from end_year

In [151]:
# fill nan with zero
#db_clean['End_Year'] = db_clean['End_Year'].fillna(0)


In [152]:
#db_clean['End_Year'] = db_clean['End_Year'].astype(int)

In [153]:
# verification
unique_year_clean2 = db_clean['End_Year'].unique().tolist()
print(unique_year_clean2)

[1969.0, nan, 1976.0, 1970.0, 1971.0, 1974.0, 1975.0, 1963.0, 1961.0, 1978.0, 1962.0, 1985.0, 1986.0, 1983.0, 1965.0, 1964.0, 1940.0, 1941.0, 1955.0, 1989.0, 1950.0, 1980.0, 1981.0, 1979.0, 1966.0, 1977.0, 1993.0, 1996.0, 1999.0, 2000.0, 1998.0, 1997.0, 1994.0, 1984.0, 1987.0, 1990.0, 1992.0, 1952.0, 1988.0, 1982.0, 1959.0, 1967.0, 1968.0, 1953.0, 1995.0, 1936.0, 1960.0, 1956.0, 1957.0, 2003.0, 1951.0, 1954.0, 2001.0, 1934.0, 1938.0, 1973.0, 2006.0, 2009.0, 2004.0, 2010.0, 1972.0, 1958.0, 1939.0, 1991.0, 1949.0, 1931.0, 1932.0, 1944.0, 1933.0, 1937.0, 1935.0, 2002.0, 2018.0, 2017.0, 1948.0, 2005.0, 2007.0]


### Cars at least 30y old

In [154]:
db_clean2 = db_clean.copy()

In [155]:
db_clean2['Start_Year'] = db_clean2['Start_Year'].fillna(0)
db_clean2['Start_Year'] = db_clean2['Start_Year'].astype(int)
db_clean2['End_Year'] = db_clean2['End_Year'].fillna(0)
db_clean2['End_Year'] = db_clean2['End_Year'].astype(int)

In [156]:
db_clean2 = db_clean2[(db_clean2['Start_Year'] <= 1994)]

In [157]:
db_clean2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1964 entries, 5 to 2271
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Brand       1964 non-null   object
 1   Model       1963 non-null   object
 2   Submodel    1698 non-null   object
 3   Origin      1955 non-null   object
 4   Start_Year  1964 non-null   int32 
 5   End_Year    1964 non-null   int32 
dtypes: int32(2), object(4)
memory usage: 92.1+ KB


### 5.3. Origin variable
<a id='5.3'>

#### Transfomation to uppercase

In [158]:
# convert all in uppercase
db_clean['Origin'] = db_clean['Origin'].str.upper()

db_clean.tail()


Brand   Model   Submodel   Origin  Start_Year  End_Year
2267     Volvo   P1800         ES   SWEDEN        1971    1973.0
2268     Volvo  P 1800          S   SWEDEN        1968    1969.0
2269     Volvo  P 1800          S   SWEDEN        1963    1968.0
2270  Wartburg     353  W Tourist  GERMANY        1975    1989.0
2271   Zundapp   Janus        NaN  GERMANY        1957    1958.0

#### replacement:  Zweden -> Sweden 

In [159]:
db_clean['Origin'] = db_clean['Origin'].replace('ZWEDEN', 'SWEDEN')


#### investigation of 1977 case

In [160]:
registo_1977 = db_clean[db_clean['Origin'] == '1977']
print(registo_1977)


    Brand Model Submodel Origin  Start_Year  End_Year
405   BMW   3.0        S   1977        1971       NaN


<span style='color:blue'> 
    
    
- it has no origin, years fly between columns of origin and start year
    
- through investigation, classic car was manufactured in Germany
    
- year and origin replacement 
    

In [161]:
# Year and origin replacement
db_clean.loc[405, 'Start_Year'] = '1971'
db_clean.loc[405, 'End_Year'] = '1977'
db_clean.loc[405, 'Origin'] = 'GERMANY'

C:\Users\InesC\AppData\Local\Temp\ipykernel_11696\2284759304.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '1971' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  db_clean.loc[405, 'Start_Year'] = '1971'
C:\Users\InesC\AppData\Local\Temp\ipykernel_11696\2284759304.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '1977' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  db_clean.loc[405, 'End_Year'] = '1977'


In [162]:
# verification
db_clean.loc[[405]]

Brand Model Submodel   Origin Start_Year End_Year
405   BMW   3.0        S  GERMANY       1971     1977

In [163]:
unique_origin_clean = db_clean['Origin'].unique().tolist()
print(unique_origin_clean)

['ITALY', 'GERMANY', 'UNITED KINGDOM', 'FRANCE', 'GREECE', 'UNITED STATES', 'AUSTRALIA', 'SPAIN', nan, 'NETHERLANDS', 'JAPAN', 'IRELAND', 'GERMANY / UNITED KINGDOM', 'GERMANY / UNITED KINGDOM UP TO 1976', 'SWEDEN']


In [164]:
filtered_db = db_clean.loc[db_clean['Origin'] == 'GERMANY / UNITED KINGDOM UP TO 1976']
filtered_db

Brand    Model             Submodel                               Origin  \
1348  Ford  Granada               2.6 V6  GERMANY / UNITED KINGDOM UP TO 1976   
1350  Ford  Granada          2.6 V6 Ghia  GERMANY / UNITED KINGDOM UP TO 1976   
1352  Ford  Granada           2.6 V6 GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1354  Ford  Granada            2.6 V6 XL  GERMANY / UNITED KINGDOM UP TO 1976   
1356  Ford  Granada               3.0 V6  GERMANY / UNITED KINGDOM UP TO 1976   
1358  Ford  Granada          3.0 V6 Ghia  GERMANY / UNITED KINGDOM UP TO 1976   
1360  Ford  Granada           3.0 V6 GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1362  Ford  Granada            3.0 V6 XL  GERMANY / UNITED KINGDOM UP TO 1976   
1363  Ford  Granada         Coupe 2.3 V6  GERMANY / UNITED KINGDOM UP TO 1976   
1364  Ford  Granada     Coupe 2.3 V6 GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1365  Ford  Granada    Coupe 2.3 V6 Ghia  GERMANY / UNITED KINGDOM UP TO 1976   
1366  Ford  Granada     Coupe 2.3 V6 GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1367  Ford  Granada      Coupe 2.3 V6 XL  GERMANY / UNITED KINGDOM UP TO 1976   
1368  Ford  Granada         Coupe 2.6 V6  GERMANY / UNITED KINGDOM UP TO 1976   
1369  Ford  Granada    Coupe 2.6 V6 Ghia  GERMANY / UNITED KINGDOM UP TO 1976   
1370  Ford  Granada     Coupe 2.6 V6 GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1371  Ford  Granada     Coupe 2.6 V6 GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1372  Ford  Granada      Coupe 2.6 V6 XL  GERMANY / UNITED KINGDOM UP TO 1976   
1373  Ford  Granada         Coupe 3.0 V6  GERMANY / UNITED KINGDOM UP TO 1976   
1374  Ford  Granada    Coupe 3.0 V6 Ghia  GERMANY / UNITED KINGDOM UP TO 1976   
1375  Ford  Granada     Coupe 3.0 V6 GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1376  Ford  Granada    Coupe 3.0 V6  GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1377  Ford  Granada      Coupe 3.0 V6 XL  GERMANY / UNITED KINGDOM UP TO 1976   
1378  Ford  Granada       Turnier 2.3 V6  GERMANY / UNITED KINGDOM UP TO 1976   
1379  Ford  Granada  Turnier 2.3 V6 Ghia  GERMANY / UNITED KINGDOM UP TO 1976   
1380  Ford  Granada   Turnier 2.3 V6 GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1381  Ford  Granada    Turnier 2.3 V6 XL  GERMANY / UNITED KINGDOM UP TO 1976   
1382  Ford  Granada       Turnier 3.0 V6  GERMANY / UNITED KINGDOM UP TO 1976   
1383  Ford  Granada  Turnier 3.0 V6 Ghia  GERMANY / UNITED KINGDOM UP TO 1976   
1384  Ford  Granada   Turnier 3.0 V6 GXL  GERMANY / UNITED KINGDOM UP TO 1976   
1385  Ford  Granada    Turnier 3.0 V6 XL  GERMANY / UNITED KINGDOM UP TO 1976   

     Start_Year End_Year  
1348       1972   1974.0  
1350       1974   1975.0  
1352       1972   1975.0  
1354       1974   1975.0  
1356       1972   1974.0  
1358       1974   1975.0  
1360       1972   1975.0  
1362       1974   1975.0  
1363       1972   1974.0  
1364       1974   1975.0  
1365       1974   1975.0  
1366       1972   1974.0  
1367       1974   1975.0  
1368       1972   1974.0  
1369       1974   1975.0  
1370       1974   1975.0  
1371       1972   1974.0  
1372       1974   1975.0  
1373       1972   1974.0  
1374       1974   1975.0  
1375       1974   1975.0  
1376       1972   1974.0  
1377       1974   1975.0  
1378       1972   1975.0  
1379       1974   1975.0  
1380       1972   1975.0  
1381       1974   1975.0  
1382       1972   1975.0  
1383       1974   1975.0  
1384       1972   1975.0  
1385       1974   1975.0

In [165]:
db_clean['Origin'] = db_clean['Origin'].replace('GERMANY / UNITED KINGDOM', 'GERMANY')

In [166]:
db_clean['Origin'] = db_clean['Origin'].replace('GERMANY / UNITED KINGDOM UP TO 1976', 'GERMANY')

In [167]:
unique_origin_clean = db_clean['Origin'].unique().tolist()
print(unique_origin_clean)

['ITALY', 'GERMANY', 'UNITED KINGDOM', 'FRANCE', 'GREECE', 'UNITED STATES', 'AUSTRALIA', 'SPAIN', nan, 'NETHERLANDS', 'JAPAN', 'IRELAND', 'SWEDEN']


In [168]:
db_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2267 entries, 5 to 2271
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Brand       2267 non-null   object
 1   Model       2266 non-null   object
 2   Submodel    2000 non-null   object
 3   Origin      2258 non-null   object
 4   Start_Year  2267 non-null   object
 5   End_Year    2137 non-null   object
dtypes: object(6)
memory usage: 106.4+ KB


In [169]:
db_clean.describe(include='all')

Brand Model Submodel Origin  Start_Year  End_Year
count      2267  2266     2000   2258        2267    2137.0
unique       85   479     1338     12          74      77.0
top     Citroen    CX       GT  ITALY        1994    1975.0
freq        336    97       23    781         102     112.0

## 6. Data exportation 
<a id='6'>

In [170]:
db_clean.head()

Brand     Model Submodel Origin Start_Year End_Year
5  Ferrari  Dino 206       GT  ITALY       1968   1969.0
6  Ferrari       206       SP  ITALY       1966      NaN
7  Ferrari       208      GT4  ITALY       1975   1976.0
8  Ferrari  Dino 246       GT  ITALY       1969   1970.0
9  Ferrari  Dino 246       GT  ITALY       1970   1971.0

In [171]:
db_clean.to_csv('Manufacturers_processeddata.csv', index=False)